drivers: ieee802154: rf2xx: Move power table to devicetree and add Sub-Giga support

nandojve opened the pull request:
The current version of power table is hardcoded in the driver which is a problem when use devices in production. This change remove all hardcode from driver and re-implement the feature to allow people create a table which is defined in devicetree. The big advantage is that each board can define their own table based on lab tests and allows use of FEM devices inclusive.

In addition, add support to at86rf212[b] sub-giga devices. This work enables use of pages 0, 2 and 5 in accordance with IEEE-802.15.4/2003/2006/2011. The proprietary speeds can be object of future work.

CC: @markus-becker-tridonic-com @maxmclau

How to build and test with AVR-RZ600 modules on Atmel Boards

```
west build -b sam4e_xpro samples/net/sockets/echo_client -- -DSHIELD=atmel_rf2xx_legacy -DOVERLAY_CONFIG=overlay-802154-subg.conf
west build -b sam4s_xplained samples/net/sockets/echo_server -- -DSHIELD=atmel_rf2xx_xplained -DOVERLAY_CONFIG=overlay-802154-subg.conf
```

Note: Ethernet/MDIO drivers should be disabled at SAM4E-XPRO

nandojve committed:
drivers: ieee802154: rf2xx: Move power table to devicetree

The current version of power table is hardcoded in the driver which is a
problem when use devices in production. This change remove all hardcode
from driver and reimplement the feature to allow people create a table
which is defined in devicetree. The big advantage is that each board can
define their own table based on lab tests and allows use of FEM devices
inclusive.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

nandojve committed:
drivers: ieee802154: rf2xx: Add support to Sub-Giga

Add support to at86rf212[b] sub-giga devices. This work enables use of
pages 0, 2 and 5 in accordance with IEEE-802.15.4/2003/2006/2011. The
proprietary speeds can be object of future work.

Note: It is recommended that user define a power table for better
performance, low emissions and to save power. A reference power table
can be found in the datasheet and should be used for tests only and
not on a final product.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

galak commented on dts/bindings/ieee802154/atmel,rf2xx.yaml:
```diff
@@ -38,3 +38,53 @@ properties:
+        This property must be used with tx-pwr-mapping. Note that this allows
+        use of external transceiver. The value is represented by two values where
+        first represent signal indication and second is the value itself. The
+        default value is set equal to tx-pwr-max to create a table with one entry.
```
which table has one entry in it?  The last sentence is confusing.

nandojve replied:
Add a more detailed description.

galak commented on dts/bindings/ieee802154/atmel,rf2xx.yaml:
```diff
@@ -38,3 +38,53 @@ properties:
+        This property must be used with tx-pwr-mapping. Note that this allows
+        use of external transceiver. The value is represented by two values where
+        first represent signal indication and second is the value itself. The
+        default value is set equal to tx-pwr-min to create a table with one entry.
```
same comment as above.

nandojve replied:
idem

galak commented on dts/bindings/ieee802154/atmel,rf2xx.yaml:
```diff
@@ -38,3 +38,53 @@ properties:
+        - 0
+        - 2
+        - 5
+      default: 0
```
since you are using DT_*PROP_OR you should remove the defaults from the binding.

nandojve replied:
Fixed

nandojve commented:
Hi @galak ,

I was wondering if you have few time to revisit this.

mbolivar-nordic commented on dts/bindings/ieee802154/atmel,rf2xx.yaml:
```diff
@@ -38,3 +38,75 @@ properties:
+        regulations. By default this value set an output power above 0dBm for all
+        transceivers. This property must be used with tx-pwr-min and tx-pwr-max
+        for normal operations. The number of elements in the array is defined by
+        CONFIG_IEEE802154_RF2XX_TX_PWR_MAP_ENTRIES Kconfig property. The max
```
This introduces a dependency from Kconfig into DT, which is a bit strange, no? Why is this necessary?

nandojve replied:
My initial thought was to define this parameter as the basis for creating and validating the size of the table. As the drive allows the use of more than one radio, the table size is shared between them and it must be the same for both instances, as this parameter is in the device's configuration structure, I don't know if it's possible to get the size directly from the size array when there are two instances because it could have different values between the two devices.

mbolivar-nordic replied:
Can you do a BUILD_ASSERT that the table sizes are all the same within a DT_INST_FOREACH_STATUS_OKAY macro? Something like this:

```c
#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) > 1
#define THE_TABLE_SIZE(inst) ARRAY_SIZE(....(inst))
#define CHECK_TABLE_SIZE(inst) BUILD_ASSERT(THE_TABLE_SIZE(inst) == THE_TABLE_SIZE(0));
DT_INST_FOREACH_STATUS_OKAY(CHECK_TABLE_SIZE)
#endif
```

galak reviewed (dismissed):
Agree with @mbolivar-nordic that having a dependency between dts and Kconfig as specified in the binding doesn't make sense.  We need to fix/clean that up.

nandojve commented:
> Agree with @mbolivar-nordic that having a dependency between dts and Kconfig as specified in the binding doesn't make sense. We need to fix/clean that up.

Implemented check as suggested by @mbolivar-nordic and dropped Kconfig option.

mbolivar-nordic reviewed (dismissed):
This review is for the DT binding only. Thanks for taking the time to make the change and for writing such great descriptions.

nandojve commented:
@carlescufi , @galak , could you revisit this? I already got @mbolivar-nordic approval on devicetree.

nandojve commented:
I was wondering if anyone could help with this PR.

nandojve commented:
Fix access to table order and rebase.

gmarull commented on drivers/ieee802154/ieee802154_rf2xx.c:
```diff
@@ -66,47 +67,15 @@ static struct net_pkt rf2xx_ack_pkt = {
-	/* 0x1, */               /*  3.7 dBm */
-	0x0                      /*  4   dBm: 4 */
-};
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) > 1
+#define THE_TABLE_SIZE(inst) DT_INST_PROP_LEN(inst, tx_pwr_table)
+#define CHECK_TABLE_SIZE(inst) \
+	BUILD_ASSERT(THE_TABLE_SIZE(inst) == TX_PWR_TABLE_SIZE, \
+		     "The TX PWR TABLE must be same size for all instances.");
+DT_INST_FOREACH_STATUS_OKAY(CHECK_TABLE_SIZE)
```
note: you can allow multiple sizes by just defining the table outside of driver's config/data, and just take a pointer to the table + its size instead. See for example https://github.com/teslabs/spinner/blob/5591731a547a9d5ee50636360122e53625cacb4b/drivers/svpwm/svpwm_stm32.c#L298-L299

gmarull reviewed (commented):
one minor nitpick

nandojve commented:
Implemented suggestion from @gmarull to have different table sizes.

nandojve commented:
Hi @mbolivar-nordic , @rlubos, @gmarull I was wondering if you could re approve PR.
Thank you!
